### DISCUSSION TITLE
Use different edges in VRP

### DISCUSSION THREAD

**Discussion by mkrechetov** [google/or-tools#3065] [Routing (and legacy CP) questions]
Good day!

I have a topology graph of, say, a roads in a country.

Is it possible to force all vehicles to use different edges of my topology graph? Could not find anything like this in the guide.

Moreover, I want vehicles to follow the edges of underlying topology graph. For example, a vehicle can travel from vertex 1 to vertex 2 only if there is an edge (1,2) in the topology graph. Otherwise, that vehicle should visit all vertices along a path from 1 to 2.

Is it possible to achieve this with Routing solver?

Thanks in advance!

**Comment by jmarca** [comment]
Do you mean solve an edge covering problem?  Depending on the details, you might be able to formulate the problem as a Chinese Postman Problem (https://en.wikipedia.org/wiki/Route_inspection_problem).  If so, and if there are no special cases, you might be able to use one of the polynomial time algorithms for that problem.  

But if you have any "interesting" constraints, it probably won't fit the polynomial time algorithm.  In that case you can transform the edge graph into its dual, and then each edge becomes a node, and it will fit the needs of this solver...visiting a node in the dual == traversing an edge in the original.

I did that once for a street sweeping problem and it worked well.

**Reply by mkrechetov** [reply]
Thanks for the reference!

In fact, I am trying to approach the following problem:

Let G be an undirected graph. Every edge has two weights: Length and Cost. I consider a pair of vertices (s,t) and want the shortest pair of path p1, p2 with the following property:

|Cost(p1)-Cost(p2)| < threshold, and
P1 and p2 do not have common edges.

By shortest pair I mean that the total Length of both paths is minized over all feasible pairs of paths between s and t.

My idea was to try feeding this problem to the Routing solver by using pickup and delivery together with cumulative constraint on Cost (so that Cost on the way to pickup equals the cost on the way to delivery point)

However, I was not able to configure the solver to follow the edges of my graph and, moreover, to use different edges on the way back.

The difference with Chinese Postman is that I do no need to cover all the edges, only to find two paths.

**Comment by jmarca** [comment]
Interesting.

of course it may not be possible to use entirely disparate paths.
There is probably some sort of network algorithm that will tell you if
that is the case for your network.  Min cut max flow, cliques, etc etc.
It's been a while since grad school.

Regardless, it is still necessary to convert the regular network into
its line graph dual.  You can still use the regular network to get the
shortest paths, etc, but you use the dual to track visiting edges.

On the other hand, now that I think about it, do you even need a
solver?

I vaguely recall using min cut to solve this kind of thing.  If min
cut between a and b has one edge, you're done, it fails.  If not, then
find the shortest path from A to B, remove the edge in the min cut set
that is on that path, then find the next shortest path from A to B
without using that critical link.

Recurse back to A, and from removed links to B

Maybe

James

On Wed, Jan 12, 2022 at 12:44:25AM -0800, mkrechetov wrote:
> Thanks for the reference!
>
> In fact, I am trying to approach the following problem:
>
> Let G be an undirected graph. Every edge has two weights: Length and Cost. I consider a pair of vertices (s,t) and want the shortest pair of path p1, p2 with the following property:
>
> |Cost(p1)-Cost(p2)| < threshold, and
> P1 and p2 do not have common edges.
>
> By shortest pair I mean that the total Length of both paths is minized over all feasible pairs of paths between s and t.
>
> My idea was to try feeding this problem to the Routing solver by using pickup and delivery together with cumulative constraint on Cost (so that Cost on the way to pickup equals the cost on the way to delivery point)
>
> However, I was not able to configure the solver to follow the edges of my graph and, moreover, to use different edges on the way back.
>
> The difference with Chinese Postman is that I do no need to cover all the edges, only to find two paths.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3065#discussioncomment-1953124
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC
